### DISCUSSION TITLE
I want to control the deployment sequence of components in the application through workflow. How do I design this workflow?

### DISCUSSION THREAD

**Discussion by lclovehoro** [kubevela/kubevela#2571] [Q&A]
I want to controll components in the application,I created the following workflow “apply-component”：
`cue:
  template: |
    import (
	  "vela/op"
	)
	
	load: op.#Load & {}
	apply: op.#ApplyComponent & {
	  value: { load.value[parameter.name] }
	}
	
	resource: op.#Read & {
	  value: {
	    Kind:	"Deployment"
		apiVersion:	"apps/v1"
		metadata: {
			name:	parameter.name
			namespace:	context.namespace
		}
	  }
	}
	
	workload: resource.value
	wait:	op.#ConditionalWait & {
	  continue: workload.status.readyReplicas == workload.status.replicas && workload.status.observedGeneration == workload.metadata.generation
	}
	
	parameter: {
	  name: string
	}`

When I do not use trait "rollout", I can achieve my purpose normally。

`apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: test
  namespace: test
spec:
  components:
  - name: nginx-1
  ...
  - name: nginx-2
  ...
  - name: nginx-3
  workflow:
    steps:
	  - name: apply-spec-component
	    type: apply-component
		properties:
		  name: nginx-2
	  - name: apply-all
	    type: apply-application`

However, when I use trait "rollout", the App will remain in the "runningWorkflow" state after publishing the specified components.
I think that after using rollout, the name of the component changed in the deployment, from "nginx-2" to "nginx-2-v1" or "nginx-2-v2", resulting in "op.#Read" can't get the true value, so I want to know about how to design a better workflow to realize the control component when using rollout.
Regarding "vela/op", can provide more details on the official website?

**Comment by leejanee** [comment]
When component use "rollout" trait, it means that the component workload `deployment` will be hosted by `rollout controller`，workflow will not apply the `deployment ` CR.
Now the rollout function is being refactored. You will see the new `rollout` realization in version 1.2(KubeVela).
